lib/cli/src/upgrade.ts

```diff
@@ -0,0 +1,105 @@
+import { sync as spawnSync } from 'cross-spawn';
+import semver from '@storybook/semver';
+import { logger } from '@storybook/node-logger';
+import { JsPackageManagerFactory } from './js-package-manager';
+import { commandLog } from './helpers';
+
+type Package = {
+  package: string;
+  version: string;
+};
+
+const versionRegex = /(@storybook\/[^@]+)@(\S+)/;
+export const getStorybookVersion = (line: string) => {
+  if (line.startsWith('npm ')) return null;
+  const match = versionRegex.exec(line);
+  if (!match || !semver.clean(match[2])) return null;
+  return {
+    package: match[1],
+    version: match[2],
+  };
+};
+
+const excludeList = [
+  '@storybook/linter-config',
+  '@storybook/design-system',
+  '@storybook/ember-cli-storybook',
+  '@storybook/semver',
+  '@storybook/eslint-config-storybook',
+  '@storybook/addon-console',
+  '@storybook/csf',
+];
+export const isCorePackage = (pkg: string) =>
+  pkg.startsWith('@storybook/') &&
+  !pkg.startsWith('@storybook/preset-') &&
+  !excludeList.includes(pkg);
+
+const deprecatedList = [
+  '@storybook/addon-notes',
+  '@storybook/addon-info',
+  '@storybook/addon-contexts',
+  '@storybook/addon-options',
+  '@storybook/addon-centered',
+];
+export const isDeprecatedPackage = (pkg: string) => deprecatedList.includes(pkg);
+
+const formatPackage = (pkg: Package) => `${pkg.package}@${pkg.version}`;
+
+const warnPackages = (pkgs: Package[]) =>
+  pkgs.forEach((pkg) => logger.warn(`- ${formatPackage(pkg)}`));
+
+export const checkVersionConsistency = () => {
+  const lines = spawnSync('npm', ['ls'], { stdio: 'pipe' }).output.toString().split('\n');
+  const storybookPackages = lines
+    .map(getStorybookVersion)
+    .filter(Boolean)
+    .filter((pkg) => isCorePackage(pkg.package));
+  if (!storybookPackages.length) {
+    throw new Error('No storybook core packages found!');
+  }
+  storybookPackages.sort((a, b) => semver.rcompare(a.version, b.version));
+  const latestVersion = storybookPackages[0].version;
+  const outdated = storybookPackages.filter((pkg) => pkg.version !== latestVersion);
+  if (outdated.length > 0) {
+    logger.warn(
+      `Found ${outdated.length} outdated packages (relative to '${formatPackage(
+        storybookPackages[0]
+      )}')`
+    );
+    logger.warn('Please make sure your packages are updated to ensure a consistent experience.');
+    warnPackages(outdated);
+  }
+
+  if (semver.gt(latestVersion, '5.4.0')) {
```

Comment:
Would we have to update this magic constant every release? Should we not pull it from `package.json`?

Reply:
Made this more general, but maybe not in the way you had in mind